hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/replication/health/RatisUnhealthyReplicationCheckHandler.java

```diff
@@ -0,0 +1,193 @@
+      LOG.info("No UNHEALTHY replicas are present for container {} with " +
+          "replicas [{}].", container, request.getContainerReplicas());
+      return false;
+    } else {
+      LOG.debug("Container {} has UNHEALTHY replicas. Checking its " +
+          "replication status.", container);
+      report.incrementAndSample(ReplicationManagerReport.HealthState.UNHEALTHY,
+          container.containerID());
+    }
+
+    ContainerHealthResult health = checkReplication(request);
+    if (health.getHealthState()
+        == ContainerHealthResult.HealthState.UNDER_REPLICATED) {
+      ContainerHealthResult.UnderReplicatedHealthResult underHealth
+          = ((ContainerHealthResult.UnderReplicatedHealthResult) health);
+      report.incrementAndSample(
+          ReplicationManagerReport.HealthState.UNDER_REPLICATED,
+          container.containerID());
+      LOG.debug("Container {} is Under Replicated. isReplicatedOkAfterPending" +
+              " is [{}]. isUnrecoverable is [{}]. hasHealthyReplicas is [{}].",
+          container,
+          underHealth.isReplicatedOkAfterPending(),
+          underHealth.isUnrecoverable(), underHealth.hasHealthyReplicas());
+
+      if (!underHealth.isReplicatedOkAfterPending()) {
+        request.getReplicationQueue().enqueue(underHealth);
+      }
+      return true;
+    }
+
+    if (health.getHealthState()
+        == ContainerHealthResult.HealthState.OVER_REPLICATED) {
+      report.incrementAndSample(
+          ReplicationManagerReport.HealthState.OVER_REPLICATED,
+          container.containerID());
+      ContainerHealthResult.OverReplicatedHealthResult overHealth
+          = ((ContainerHealthResult.OverReplicatedHealthResult) health);
+      LOG.debug("Container {} is Over Replicated. isReplicatedOkAfterPending" +
+              " is [{}]. hasMismatchedReplicas is [{}]", container,
+          overHealth.isReplicatedOkAfterPending(),
+          overHealth.hasMismatchedReplicas());
+
+      if (!overHealth.isReplicatedOkAfterPending()) {
+        request.getReplicationQueue().enqueue(overHealth);
+      }
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   * Verify there's no under or over replication if only healthy replicas are
+   * considered, or that there are no healthy replicas.
+   * @return true if there's no under/over replication considering healthy
+   * replicas
+   */
+  private boolean verifyPerfectReplication(ContainerCheckRequest request) {
+    RatisContainerReplicaCount replicaCountWithoutUnhealthy =
+        new RatisContainerReplicaCount(request.getContainerInfo(),
+            request.getContainerReplicas(), request.getPendingOps(),
+            request.getMaintenanceRedundancy(), false);
+
+    if (replicaCountWithoutUnhealthy.getHealthyReplicaCount() == 0) {
+      return true;
+    }
+    if (!replicaCountWithoutUnhealthy.isSufficientlyReplicated() ||
+        replicaCountWithoutUnhealthy.isOverReplicated()) {
```

Comment:
This check looks correct although it is confusing. Sufficiently replicated means there are >= 0 excess replicas, and over replicated means there are > 0 excess replicas. So this says `if !(excess >= 0) || excess > 0` which is `if excess < 0 || excess > 0`. So this does check that there is not perfect replication, but in a roundabout way. The `isSufficientlyReplicated` method is confusing to me in general. It seems it could be replaced by a more explicit `isUnderReplicated` method and negating the result when needed. This check would then be more literal:
```
if (replicaCountWithoutUnhealthy.isUnderReplicated() ||
        replicaCountWithoutUnhealthy.isOverReplicated()) {
```

Reply:
Adding an `isUnderReplicated` method here to make it more readable. It will use `isSufficientlyReplicated` internally. I'm averse to completely replacing it because there are probably some places where `isSufficientlyReplicated` reads better. We can revisit this in another jira.